collector/src/etw_parser.rs

```diff
@@ -0,0 +1,677 @@
+                    thread_id: columns[pmc_thread_id].trim().parse()?,
+                    instructions_retired: columns[pmc_instructions_retired].trim().parse()?,
+                    total_cycles: columns[pmc_total_cycles].trim().parse()?,
+                });
+            }
+            CSWITCH => {
+                let timestamp = columns[cswitch_timestamp].trim().parse()?;
+                let old_process = columns[cswitch_old_process].trim();
+                let old_pid = extract_pid(old_process);
+                let new_process = columns[cswitch_new_process].trim();
+                let new_pid = extract_pid(new_process);
+
+                if !currently_watched_processes.contains(&old_pid) && !currently_watched_processes.contains(&new_pid) {
+                    // In this case, the previous Pmc event at this same timestamp isn't relevant.
+                    // There might not be a previous event if the CSwitch event occurs before the
+                    // Pmc events start recording.
+                    if let Some(pmc) = &last_pmc {
+                        assert_eq!(timestamp, pmc.timestamp);
+                        last_pmc = None;
+                    }
+
+                    continue;
+                }
+
+                events.push((std::mem::take(&mut last_pmc).unwrap(), CSwitch {
+                    timestamp,
+                    old_process_pid: old_pid,
+                    new_process_pid: new_pid,
+                    cpu: columns[cswitch_cpu].trim().parse()?,
+                }));
+            }
+            _ => { }
+        }
+    }
+
+    Ok(EventData {
+        events,
+        watched_processes: all_watched_processes,
+    })
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Counters {
+    pub instructions_retired: u64,
+    pub total_cycles: u64,
+    pub cpu_clock: f64,
+}
+
+impl std::ops::Add for Counters {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self {
+            instructions_retired: self.instructions_retired + rhs.instructions_retired,
+            total_cycles: self.total_cycles + rhs.total_cycles,
+            cpu_clock: self.cpu_clock + rhs.cpu_clock,
+        }
+    }
+}
+
+impl std::ops::Sub for Counters {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        assert!(rhs.instructions_retired < self.instructions_retired);
+        assert!(rhs.total_cycles < self.total_cycles);
+
+        Self {
+            instructions_retired: self.instructions_retired - rhs.instructions_retired,
+            total_cycles: self.total_cycles - rhs.total_cycles,
+            cpu_clock: self.cpu_clock - rhs.cpu_clock,
+        }
+    }
+}
+
+impl Default for Counters {
+    fn default() -> Self {
+        Self {
+            instructions_retired: 0,
+            total_cycles: 0,
+            // FIXME(wesleywiser): We should be properly calculating this value by taking the total time
```

Comment:
Should we create issues in the repo for these FIXMEs?

Reply:
That seems reasonable to me! I'll create issues when this merges and link to the FIXME lines.